ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java

```diff
@@ -164,6 +166,8 @@ class GetBlobOperation extends GetOperation {
         shouldLookupMetadataCache() ? (BlobMetadata) blobMetadataCache.getObject(blobId.toString()) : null;
     firstChunk = (blobMetadata == null) ? new FirstGetChunk() : new CachedFirstChunk(blobMetadata);
     this.nonBlockingRouter = nonBlockingRouter;
+    decompressionService = new CompressionService(routerConfig.getCompressionConfig(),
```

Comment:
Do we create a new `CompressionService` per `GetOperation`? Can this be created at `OperationController` and be shared by all `GetOperation`s?

Reply:
Eventually we want the compression service to be a singleton using Dependency Injection.  For this case, we should not create an instance in controller and pass in here because it breaks the encapsulation principal that the class itself should be self-contain and hold on to objects it needs to complete its operation.  Clients (operation class in this case) should not interfere or decide how the dependent class should behave.  This is also coding style thing.  That means class A should use class B by its APIs, not class A controls how class B implements its methods.  In this case, if Operation class pass in CompressionService to PutOperation, then Operation class forces compression service must be singleton, and put operation must have compression as its process.  In this pattern, classes know too much about other classes and interfere their implementation.  Instead, classes should be clean, self-contained, and abstracted so that they communicate only by API without knowing how the classes are implemented.